iml-manager-cli/src/diff.rs

```diff
@@ -0,0 +1,222 @@
+    let n = right_xs.len();
+    let mut i = 0; // i ∈ (0..m)
+    let mut j = 0; // j ∈ (0..n)
+    let mut result = Vec::with_capacity(m + n);
+    loop {
+        if i >= m && j >= n {
+            // both arrays exhausted
+            return result;
+        } else if j >= n {
+            // right array exhausted
+            (i..m).for_each(|i| result.push(AlignmentOp::Insert(Side::Right, i, j)));
+            return result;
+        } else if i >= m {
+            // left array exhausted
+            (j..n).for_each(|j| result.push(AlignmentOp::Insert(Side::Left, i, j)));
+            return result;
+        } else if left_xs[i] == right_xs[j] {
+            // both elements are the same, skip
+            i += 1;
+            j += 1;
+        } else if left_xs[i].key() == right_xs[j].key() {
+            result.push(AlignmentOp::Replace(Side::Left, i, j));
+            i += 1;
+            j += 1;
+        } else {
+            let l = &left_xs[i];
+            if let Some((j1, _)) = right_xs[j..]
+                .iter()
+                .enumerate()
+                .find(|(_, r)| r.key() == l.key())
+            {
+                (0..j1).for_each(|ix| result.push(AlignmentOp::Insert(Side::Left, i, j + ix)));
+                j += j1;
+            } else if let Some((_, _)) = right_xs[0..j]
+                .iter()
+                .enumerate()
+                .rfind(|(_, r)| r.key() == l.key())
+            {
+                result.push(AlignmentOp::Delete(Side::Left, i));
+                i += 1;
+            } else {
+                result.push(AlignmentOp::Insert(Side::Right, i, j));
+                i += 1;
+            }
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use indicatif::{MultiProgress, ProgressBar, ProgressStyle};
+    use std::collections::HashMap;
+    use std::sync::Arc;
+    use std::thread;
+
+    #[derive(Debug, Copy, Clone, PartialEq, Eq, Hash)]
+    struct Id(i32);
+
+    #[derive(Debug, Clone, PartialEq, Eq)]
+    struct T {
+        id: Id,
+        msg: String,
+    }
+
+    impl Keyed for T {
+        type Key = Id;
+
+        fn key(&self) -> Self::Key {
+            self.id
+        }
+    }
+
+    #[test]
+    fn test_calculate_diff() {
+        let mut xs = vec![t(1), t(2), t(3), t(10), t(11)];
+        let mut ys = vec![t(1), t(3), t(2), t(10), t(11), t(12), t(20), t(30)];
+        let diff = calculate_diff(&xs, &ys);
+        apply_diff(&mut xs, &mut ys, &diff);
+        assert_eq!(xs, ys);
+    }
+
+    #[test]
+    fn test_empty() {
+        let mut xs: Vec<T> = vec![];
+        let mut ys: Vec<T> = vec![];
+        let diff = calculate_diff(&xs, &ys);
+        apply_diff(&mut xs, &mut ys, &diff);
+        assert_eq!(xs, ys);
+    }
+
+    #[test]
+    fn test_empty_left() {
+        let mut xs = vec![];
+        let mut ys = vec![t(1), t(2), t(3), t(10), t(11)];
+        let diff = calculate_diff(&xs, &ys);
+        apply_diff(&mut xs, &mut ys, &diff);
```

Comment:
This doesn't make the test tautological! While I agree, that there should be some assertions on the `diff` values, such tests _not_ equivalent to no-op.
E.g. imagine you have serialization/deserialization functions. Two important tests are:
```rust
loop {
  let obj = // generate random data;
  let obj2 = deser(ser(obj)).unwrap();
  assert_eq!(obj, obj2);
}
```
and
```rust
loop {
  let str = // generate random string, but with taking into account the format
  match deser(str) {
    Ok(obj) => assert_eq!(str, ser(obj)),
    Err(_) => {}
  }
}
```
You also may notice, that the tests are easily fooled (hence we need to some additional tests for the format itself), nevertheless they are checking the important invariants.

Reply:
You are right, but the first test is tautological, this is why it can be easily fooled :) Here is why:

Given two functions `ser` and `deser` (which must have some value or else why do they exist?) we must have at least two of three kinds of tests: ser+deser, ser+deser(ser), deser+ser(deser), connecting three points: `A -ser-> B -deser-> C`.

You mentioned the second option. But the best one is the first, because in all other setups, one test is warranted by another, i. e. it depends on another to be correct. While we want unit tests to be independent of each other.

This is somehow mitigated in Rust because `ser` must return completely different type and Rust is not lazily cutting corners. In general, `deser(ser(obj))` can be optimized down to just `obj` (you should know it from Haskell).  We may argue that this is not the case in our implementation, but relying on the implementation (including language and compiler features) is another sign of tautology.

And in the of case of the `calculate_diff` tests, we never use `apply_diff` except in the tests, it has no value at all (yet?). We could have much smaller and more reliable code, if you just asserted diffs, not objects.

You added more tests to fix the old ones, but in fact you can now remove the old ones (and `apply_diff` as well).

Reply:
Your argument is not flawless, I'll elaborate later :-)